ocsp/responder/multi_source.go

```diff
@@ -54,50 +66,90 @@ func (src *multiSource) Response(ctx context.Context, req *ocsp.Request) (*Respo
 		src.counter.WithLabelValues("primary_result").Inc()
-		return primaryResult.resp, primaryResult.err
+		return primaryResponse, nil
+	}
+
+	// The primary response was too stale to (ideally) serve. This will be
+	// a common path once we stop ocsp-updater from writing updated blobs
+	// to MariaDB. Try to serve from the secondary.
+	var secondaryResponse *Response
+	select {
+	case <-ctx.Done():
+		src.counter.WithLabelValues("timed_out_awaiting_secondary").Inc()
+		// Best-effort: return the primary response even though it's stale.
+		return primaryResponse, nil
 
 	case secondaryResult := <-secondaryChan:
-		// If secondary returns first, wait for primary to return for
-		// comparison.
-		var primaryResult responseResult
+		if secondaryResult.err != nil {
+			if errors.Is(secondaryResult.err, rocsp.ErrRedisNotFound) {
+				// This case will happen for several hours after first issuance.
+				src.counter.WithLabelValues("primary_stale_secondary_not_found").Inc()
+			} else {
+				src.counter.WithLabelValues("primary_stale_secondary_error").Inc()
+			}
+
+			// Best-effort: return the primary response even though it's stale.
+			return primaryResponse, nil
+		}
+		secondaryResponse = secondaryResult.resp
+	}
 
-		// Listen for cancellation or timeout waiting for primary result.
-		select {
-		case <-ctx.Done():
-			src.counter.WithLabelValues("timed_out").Inc()
-			return nil, fmt.Errorf("looking up OCSP response for serial: %s err: %w", serialString, ctx.Err())
+	// If the secondary response status doesn't match primary, return
+	// primary response. For instance this will happen for several hours
+	// after any revocation.
+	if secondaryResponse.Status != primaryResponse.Status {
+		src.counter.WithLabelValues("stale_primary_status_wins").Inc()
+		return primaryResponse, nil
+	}
 
-		case primaryResult = <-primaryChan:
-		}
+	// If the primary response is fresher than the secondary, return the
+	// primary response. If ocsp-updater is updating Redis, this shouldn't
+	// happen (since revocation cases are caught above).
+	if primaryResponse.ThisUpdate.After(secondaryResponse.ThisUpdate) {
+		src.counter.WithLabelValues("primary_newer").Inc()
+		return primaryResponse, nil
+	}
 
-		// Check for error returned from the primary lookup, return on error.
-		if primaryResult.err != nil {
-			src.counter.WithLabelValues("primary_error").Inc()
-			return nil, primaryResult.err
-		}
+	src.counter.WithLabelValues("primary_stale_secondary_success").Inc()
+	return secondaryResponse, nil
+}
 
-		// Check for error returned from the secondary lookup. If error return
-		// primary lookup result.
+// checkSecondary updates the src.counter metrics when we're planning to return
+// a primary response. It checks if the secondary result has already arrived
+// (without blocking on it) and updates the metrics accordingly.
+func (src *multiSource) checkSecondary(primaryResponse *Response, secondaryChan <-chan responseResult) {
```

Comment:
An alternative would be for this function to block on the secondary, but to be called in a goroutine that we don't wait on. That might get us more accurate notfound/error/good stats (rather than just "slow"), at the cost of extra goroutines. Not sure if it would be worth it or not.

Reply:
Right now I think making it not block is the right call; as we get more confidence with Redis, we should switch to blocking. Otherwise, slow Redis will result in slowness for OCSP overall.

Reply:
I guess I was saying "this function could block on it, but the responder would not block the web request on this function". That way the responder's latency wouldn't be affected, just its number of concurrent goroutines (because each request would spin up an additional goroutine which potentially outlives the original request).

Reply:
Sounds good!